sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,436 @@
+//
+// The hierarchical graph itself is an independent layer from the payloads
+// specific to the Runner API, for ease of reuse by e.g. the Fn API.
+message Pipeline {
+
+  // (Required) A map from pipeline-scoped id to graph node.
+  //
+  // Each node is required to contain a PTransform specification.
+  map<int64, GraphNode> transform_nodes = 1;
+
+  // (Required) A map from pipeline-scoped id to PCollection.
+  map<int64, PCollection> pcollections = 2;
+
+  // (Required) A map from pipeline-scoped id to WindowingStrategy.
+  map<int64, WindowingStrategy> windowing_strategies = 3;
+
+  // (Required) A map from pipeline-scoped id to Coder.
+  map<int64, Coder> coders = 4;
+
+  // (Required) A map from pipeline-scoped id to Environment.
+  map<int64, Environment> environments = 5;
+
+  // (Optional) Map from option name to arbitrary serialization of its value.
+  //
+  // Some options may be user-specific or SDK-specific, in which case the
+  // payload is expected to be unintelligible.
+  //
+  // Some options may be standardized across SDKs, in which case the universe
+  // of possible values is determined by the name of the pipeline option.
+  //
+  // TODO: document what is standardized
+  map<string, google.protobuf.Any> pipeline_options = 6;
+
+  // (Optional) Static display data for the pipeline.
+  map<string, string> display_data = 7;
+
+  // TODO: consider identifying the SDK + runner shim submitting the pipeline
+}
+
+// A generic node in a bipartite directed hierarchical graph.
+message GraphNode {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A URN that describes what kind graph node this is.
+  //
+  // Specifically, the URN should be enough to decipher the inputs,
+  // outputs, and payload.
+  string urn = 2;
+
+  // (Optional) if this node is contained within a composite, a pointer to the
+  // parent.
+  int64 parent_ref = 3;
+
+  // (Required) A map of local input names to target definitions.
+  //
+  // The URN for the graph node may clarify the type of the inputs
+  // (resp. outputs). For example:
+  //
+  //  - in the Runner API these are PCollections
+  //  - in the Fn API they may be Grpc ports
+  //
+  // The payload for this graph node may clarify the relationship of these
+  // inputs. For example:
+  //
+  //  - for a Flatten transform they are merged
+  //  - for a ParDo transform, some may be side inputs
+  //
+  // All inputs are recorded here so that the topological ordering of
+  // the graph is consistent whether or not the payload is understood.
+  //
+  // TODO: fix this to be int64 as they will always be refs?
+  map<string, google.protobuf.Any> inputs = 4;
+
+  // (Required) A map from local output name to PCollection definitions.
+  //
+  // The URN or payload for the graph node may clarify the type and
+  // relationship of these. For example:
+  //
+  //  - for a ParDo transform, these are tags on PCollections, which will be
```

Comment:
Suggestion grokked. Just to avoid splitting the thread on primitive (atomic or otherwise) transforms being in the proto, let's close this thread and keep it on the other.